Rubberduck.Parsing/VBA/ParseCoordinator.cs

```diff
@@ -86,17 +103,86 @@ private void ReparseRequested(object sender, EventArgs e)
                 token = _currentCancellationTokenSource.Token;
             }
 
+            BeginParse(sender,token);
+        }
+
+        public void SuspendRequested(object sender, RubberduckStatusSuspendParserEventArgs e)
+        {
+            object parseRequestor = null;
+            ParserState originalStatus;
+
+            try
+            {
+                if (_parsingSuspendLock.IsReadLockHeld)
+                {
+                    const string errorMessage =
+                        "A suspension action was attempted while a read lock was held. This indicates a bug in the code logic as suspension should not be requested from same thread that has a read lock.";
+                    Logger.Error(errorMessage);
+#if DEBUG
+                    Debug.Assert(false, errorMessage);
+#endif
+                    return;
+                }
+                if (!_parsingSuspendLock.TryEnterWriteLock(e.MillisecondsTimeout))
+                {
+                    e.TimedOut = true;
+                    return;
+                }
+
+                lock (_suspendStackSyncObject)
+                {
+                    _isSuspended = true;
+                }
+
+                originalStatus = State.Status;
+                _parserStateManager.SetStatusAndFireStateChanged(e.Requestor, ParserState.Busy,
+                    CancellationToken.None);
+                e.BusyAction.Invoke();
+            }
+            finally
+            {
+                lock (_suspendStackSyncObject)
+                {
+                    _isSuspended = false;
+                    if (_requestorStack.TryPop(out var lastRequestor))
+                    {
+                        _requestorStack.Clear();
+                        parseRequestor = lastRequestor;
+                    }
+                }
+
+                if (_parsingSuspendLock.IsWriteLockHeld)
+                {
+                    _parsingSuspendLock.ExitWriteLock();
+                }
+            }
 
-            if (!_isTestScope)
+            if (parseRequestor != null)
+            {
+                BeginParse(parseRequestor, _currentCancellationTokenSource.Token);
+            }
+            else if (originalStatus != ParserState.Ready)
             {
-                Task.Run(() => ParseAll(sender, token), token);
+                // We can't assume that we can return to any other state that's not "Ready"; it's better
+                // to request a state evaluation
+                _parserStateManager.EvaluateOverallParserState(CancellationToken.None);
             }
             else
             {
-                ParseInternal(token);
+                _parserStateManager.SetStatusAndFireStateChanged(e.Requestor, originalStatus,
```

Comment:
You're right. That is past the suspension marker. Would it be OK if we evaluated this before I clear the suspension marker? Otherwise, we'd be forced to re-parse every time a busy action was requested.

Reply:
This definitely has to happen after removing the suspension marker. Anyway, you can have a parse waiting already before the the suspension marker gets set.

Moving this in front of releasing the write lock would remove the state problem in principle. However, than any synchronous version of `BeginParse` would deadlock.

Reply:
Actually, this would only fire only if the `originalStatus` is at `Ready` and the stack is empty. In the preceding clause we do run `EvaluateOverallParserState` since we can't do anything with any other states, because that's not the suspension's concern. 

As for the inspections running - I might have to add a flag to the event arguments to indicate that we are exiting a busy state so that the inspector will ignore that case. Not clean but I'm not sure there's a better way to keep the concerns separated between the suspension and the inspections.